File: 2_method.py
class Person:  #class name 
    
    species = "Homo sapiens"  # Class attribute (shared by all instances)
    def __init__(self , name , address) -> None: #constructor 
        
        self.name = name     #attribute
        self.address = address   #attribute
        print(self.name , self.address )
        
    def intro(self):  #Instance methods
        return (f"My name is {self.name} and I live in {self.address}")



        # Class method  
    @classmethod
    def get_species(cls):
        return f"All humans are {cls.species}"
    
    #This is the static method  and A static method is a method that belongs to a class but does not access or modify the class or instance state.
    @staticmethod  
    def static_method():
        print("I am a static Method")


p = Person('Foysal Hossain MUnna' , 'Dhonmondi')  # object or instance 
p = Person('Rakib sarker' , 'Elephant Road')  # object or instance 
p = Person('Tasfia' , 'new Market')  # object or instance 
print(p.get_species())
print(p.intro())
p.static_method()






















class A:
    pass

def display():
    print("I am Methods")

obj = A()
obj.value = display

print(obj.value)
print(obj)

File: 2_method.cpp
#include <iostream>
#include <string>

class Person {
private:
    // Instance attributes (equivalent to Python's self.name and self.address)
    std::string name;
    std::string address;

public:
    // Class attribute (equivalent to Python's species, shared by all instances)
    static std::string species;

    // Constructor (equivalent to __init__)
    Person(std::string n, std::string a) : name(n), address(a) {
        std::cout << name << " " << address << std::endl;
    }

    // Instance method (equivalent to Python's intro)
    std::string intro() {
        return "My name is " + name + " and I live in " + address;
    }

    // Class method (equivalent to Python's @classmethod)
    static std::string get_species() {
        return "All humans are " + species;
    }

    // Static method (equivalent to Python's @staticmethod)
    static void static_method() {
        std::cout << "I am a static method." << std::endl;
    }
};

// Initialize static class attribute outside the class definition
std::string Person::species = "Homo sapiens";

int main() {
    // Creating objects (instances)
    Person p1("Foysal Hossain MUnna", "Dhanmondi");
    Person p2("Rakib Sarker", "Elephant Road");
    Person p3("Tasfia", "New Market");

    // Calling the class method
    std::cout << Person::get_species() << std::endl;

    // Calling the instance method
    std::cout << p1.intro() << std::endl;

    // Calling the static method
    Person::static_method();

    return 0;
}
